README.md
<h1 align="center">kdam</h1>

<p align="center">
  <a href="https://crates.io/crates/kdam">
    <img src="https://img.shields.io/crates/d/kdam?style=flat-square">
  </a>
  <a href="https://crates.io/crates/kdam">
    <img src="https://img.shields.io/crates/v/kdam?style=flat-square">
  </a>
  <a href="https://github.com/clitic/kdam">
    <img src="https://img.shields.io/github/actions/workflow/status/clitic/kdam/tests.yml?logo=github&style=flat-square">
  </a>
  <a href="https://docs.rs/kdam/latest/kdam">
    <img src="https://img.shields.io/docsrs/kdam?logo=docsdotrs&style=flat-square">
  </a>
  <a href="https://github.com/clitic/kdam#license">
    <img src="https://img.shields.io/crates/l/kdam?style=flat-square">
  </a>
  <a href="https://github.com/clitic/kdam">
    <img src="https://img.shields.io/github/repo-size/clitic/kdam?style=flat-square">
  </a>
  <a href="https://github.com/clitic/kdam">
    <img src="https://img.shields.io/tokei/lines/github/clitic/kdam?logo=github&style=flat-square">
  </a>
</p>

kdam is a console progress bar library for rust. It is port of [tqdm](https://github.com/tqdm/tqdm) library which is written in python. kdam supports all features of tqdm except few. Some features of tqdm can't be ported directly so they are implemented in different way like `RowManager` which manages multiple progress bars but in tqdm progress bars are automatically managed using `nrows`. In addition to tqdm existing features kdam also provides extra features such as spinners, charset with fill, gradient colours etc. Since kdam is written in rust its upto 4 times faster than tqdm.

Instantly make your loops show a smart progress meter. Just wrap any iterator with tqdm!(iterator) macro and you're done!

```rust
use kdam::tqdm;

fn main() {
    for _ in tqdm!(0..100) {}
}
```

```
100%|█████████████████████████████| 100/100 [00:00<00:00, 25854.49it/s]
```

kdam also supports different animation styles. All available animation styles are:

[examples/showcase/animations.rs](https://github.com/clitic/kdam/blob/main/examples/showcase/animations.rs)
![showcase_animations](https://raw.githubusercontent.com/clitic/kdam/main/images/animations.gif)

kdam also supports [rich.progress](https://rich.readthedocs.io/en/latest/progress.html) style bars with customizable columns.

[examples/rich.rs](https://github.com/clitic/kdam/blob/main/examples/rich.rs)
![showcase_rich_progress_animation](https://raw.githubusercontent.com/clitic/kdam/main/images/rich_progress.gif)

kdam doesn't restrict you to use default progress bar styles. You can create your own progress bar template using bar_format. Here is clone of [alive-progress](https://github.com/rsalmei/alive-progress) using kdam.

[examples/template.rs](https://github.com/clitic/kdam/blob/main/examples/template.rs)
![showcase_alive_progress_template](https://raw.githubusercontent.com/clitic/kdam/main/images/template.gif)

A gradient progress bar can also be created.

[examples/coloured/gradient.rs](https://github.com/clitic/kdam/blob/main/examples/coloured/gradient.rs)
![showcase_gradient](https://raw.githubusercontent.com/clitic/kdam/main/images/gradient.gif)

## Getting Started

Add this to your Cargo.toml file.

```toml
[dependencies]
kdam = "0.3.0"
```

Or add from command line.

```bash
$ cargo add kdam
```

See [docs](https://docs.rs/kdam/latest/kdam) and [examples](https://github.com/clitic/kdam/tree/main/examples) to 
know how to use it.

## Usage

### Iterator Based

```rust
use kdam::tqdm;

fn main() {
    let chars = ["a", "b", "c", "d"];
    let mut charset = String::new();

    for i in tqdm!(chars.iter()) {
        charset += i;
    }

    eprint!("\n");
    assert_eq!(charset, "abcd");
}
```

### Manual

```rust
use kdam::{tqdm, BarExt};

fn main() {
    let mut pb = tqdm!(total = 100);

    for _ in 0..100 {
        pb.update(1);
    }

    eprint!("\n");
}
```

Another example without a total value. This only shows basic stats.

```rust
use kdam::{tqdm, BarExt};

fn main() {
    let mut pb = tqdm!();

    for _ in 0..10000000 {
        pb.update(1);
    }
    pb.refresh();

    eprint!("\n");
}
```

```
10000000 [00:03, 2998660.35it/s]
```

## Examples

### Description And Additional Stats

Custom information can be displayed and updated dynamically on `kdam` bars with the `desc` and `postfix`.

```rust
use kdam::{tqdm, BarExt};

fn main() {
    let mut pb = tqdm!(total = 10);
    pb.set_postfix(format!("str={}, lst={:?}", "h", [1, 2]));
    pb.refresh();

    for i in 0..10 {
        std::thread::sleep(std::time::Duration::from_secs_f32(0.5));
        pb.set_description(format!("GEN {}", i));
        pb.update(1);
    }
    
    eprint!("\n");
}
```

```
GEN 4:  50%|█████████▎        | 5/10 [00:02<00:02, 1.95it/s, str=h, lst=[1, 2]]
```

### Nested Progress Bars

`kdam` supports nested progress bars. For manual control over positioning (e.g. for multi-processing use), you may specify `position=n` where `n=0` for the outermost bar, `n=1` for the next, and so on.

```rust
use kdam::tqdm;

fn main() {
    for _ in tqdm!(0..4, desc = "1st loop", position = 0) {
        for _ in tqdm!(0..5, desc = "2nd loop", position = 1) {
            for _ in tqdm!(0..50, desc = "3rd loop", position = 2) {
                std::thread::sleep(std::time::Duration::from_secs_f32(0.0001));
            }
        }
    }

    eprint!("{}", "\n".repeat(3));
    println!("completed!");
}
```

```
1st loop:  50%|███████▎      | 2/4 [00:08<00:08, 0.25it/s]
2nd loop:  60%|████████▌     | 3/5 [00:02<00:01, 1.25it/s]
3rd loop:   0%|▎               | 0/50 [00:00<00:00, ?it/s]
```

### Writing Messages And Inputs

Since `kdam` uses a simple printing mechanism to display progress bars, you should not write any message in the terminal using `println!()` while a progressbar is open.

To write messages in the terminal without any collision with `kdam` bar display, a `.write()` method is provided. This message will print at bar output location, which is stderr by default.

```rust
use kdam::{tqdm, BarExt};

fn main() {
    let mut pb = tqdm!(total = 10);

    for i in 0..10 {
        std::thread::sleep(std::time::Duration::from_secs_f32(0.1));

        pb.update(1);
        pb.write(format!("Done task {}", i));
    }

    eprint!("\n");
}
```

```
Done task 0
Done task 1
Done task 2
Done task 3
Done task 4
Done task 5
Done task 6
Done task 7
Done task 8
Done task 9
100%|███████████████████████████| 10/10 [00:02<00:00, 4.31it/s]
```

Similarly `.input()` method can be called to store an user input.

```rust
use kdam::{tqdm, BarExt};

fn main() {
    let mut pb = tqdm!(total = 10);

    for i in 0..10 {
        if i == 5 {
            if pb.input("Break Loop [y/n]: ").unwrap().trim() == "y" {
                break;
            }
        }

        pb.update(1);
    }

    eprint!("\n");
}
```

```
Break Loop [y/n]: y
 50%|███████████████▎              | 5/10 [00:01<00:01, 3.83it/s]
```

### Terminal Colorization

kdam also provides a text colorization trait for printing colored text in terminal. It can be used as an alternative for existing [colored](https://github.com/mackwic/colored) crate. Note that tty detection is not implemented, an alternative is to use [atty](https://github.com/softprops/atty) crate to detect if stream is tty and then use [set_colorize](https://docs.rs/kdam/latest/kdam/term/fn.set_colorize.html) function.

```rust
use kdam::term::Colorizer;

println!("{}", "hello world!".colorize("bold red"));
println!("{}", "hello world!".colorize("bright white on blue"));
```

## License

Dual Licensed

- [Apache License, Version 2.0](https://www.apache.org/licenses/LICENSE-2.0) ([LICENSE-APACHE](LICENSE-APACHE))
- [MIT license](https://opensource.org/licenses/MIT) ([LICENSE-MIT](LICENSE-MIT))


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Fixed

- Apply padding to counter.

## [0.3.0] - 2022-12-05

## Added

- `inverse_unit` field in `kdam::Bar` for displaying progress rate in inverse form.
- `set_colorize` function in `term` module to enable/disable colours.

## Changed

- Removed `prelude` module.
- Spinner support is now optional and is kept under `spinner` feature.
- `file` field is removed from `Bar` and it is replaced with `BarExt::write_to` method.
- `BarBuilder::build` methods now returns a `Result` type.

## [0.2.7] - 2022-10-11

## [0.2.6] - 2022-10-11

## [0.2.5] - 2022-10-11

### Changed

- Templating abilites are put under an feature named `template`.

### Fixed

- capacity overflow for `RichProgress` for very small window size.
- Gradient progress bar

## [0.2.4] - 2022-08-29

### Fixed

- Fixed `tqdm!(iter)` pattern according to [ISSUE #4](https://github.com/clitic/kdam/issues/4).

## [0.2.3] - 2022-08-19

### Added

- Number of columns of progress bar can now be adjusted by KDAM_NCOLS envoirnment variable.
- New `kdam::RowManager` for managing many progress bars. [ISSUE #3](https://github.com/clitic/kdam/issues/3)
- New `kdam::Bar.bar_format` field.
- New `kdam::Animation::CustomWithFill` variant.
- Better terminal colorization through `rgb(255,255,255)` and `gradient(#5A56E0,#EE6FF8)` values support.
- Now supports spinners.
    - `kdam::Spinner`
    - `tqdm!(spinner=..)`

### Changed

-  `eprint_at!` replaced with `Writer::Stderr.print_at`.
- `kdam::BarMethods` renamed to `kdam::BarExt`.
- `wrap` field removed from `kdam::Bar`.

### Fixed

- Fixed `BarMethods::clear` method for bar position above 0.
  
## [0.2.2] - 2022-07-24

### Added

- New `kdam::BarBuilder` struct.
- Implemented `From<&str>` for `kdam::Animation` and `kdam::term::Writer`.

### Changed

- All fields in `kdam::Bar` are private.
- `tqdm` macro now uses `kdam::BarBuilder` for setting values instead of setting values in `kdam::Bar` directly.
- `&str` consuming methods are switched with generic `Into<String>` trait.
- `kdam::term::get_columns` is now renamed to `kdam::term::get_columns_or`.

## [0.2.1] - 2022-07-21

### Added

- Added `reset` method `BarMethods` trait.
- New `initial` field for `kdam::Bar`.

## [0.2.0] - 2022-07-21

### Added

- Optimizations for printing mechanism.
- Optimized rich pulsating bars.

### Changed

- There are many changes in API please read [documentation](https://docs.rs/kdam/0.2.0/kdam/index.html).

### Fixed

- Fixed `Animation::Classic` animation lags.

## [0.1.7] - 2022-05-18

### Fixed

- Fixed displaying counter instead of total value.

## [0.1.6] - 2022-05-11

### Added

- Several optimizations (reduced memory consumption by 30%).
- [rich.progress](https://rich.readthedocs.io/en/latest/progress.html) style progress bar support with custom builder.
- Monitor mode is now supported. Use `kdam::monitor`.
- New `wrap` field for `kdam::Bar`.

### Changed

- `lock` module has now `AtomicBool` lock.
- `ascii` and few more fields are removed from `kdam::Bar`.
- `BarInternal` is now merged into `kdam::Bar`.
- All `u64` data types are changed to `usize`.
- Spinner removed from unbounded progress bar

### Fixed

- Unexcepted default print with fira code animation.
- `refresh` method fixed when using `max_fps = true`.
- Fixed `set_ncols` method to use full terminal length.

## [0.1.2] - 2022-04-28

### Added

- [Fira Code](https://github.com/tonsky/FiraCode) animation style i.e. `kdam::Animation::FiraCode`.
- Spinner for unknown length progress bar.
- A method to `set_position` is added to `kdam::Bar` by referring issue [#1](https://github.com/clitic/kdam/issues/1)
- Support to take input into a `String` by method `input`.

### Fixed

- `set_colour` method of `kdam::Bar` is fixed.

## [0.1.0] - 2022-03-20

[Unreleased]: https://github.com/clitic/kdam/compare/v0.3.0...HEAD
[0.3.0]: https://github.com/clitic/kdam/compare/799b34c...v0.3.0
[0.2.7]: https://github.com/clitic/kdam/compare/a206ef9...799b34c
[0.2.6]: https://github.com/clitic/kdam/compare/7b6497d...a206ef9
[0.2.5]: https://github.com/clitic/kdam/compare/fd14805...7b6497d
[0.2.4]: https://github.com/clitic/kdam/compare/970d9d9...fd14805
[0.2.3]: https://github.com/clitic/kdam/compare/15a5398...970d9d9
[0.2.2]: https://github.com/clitic/kdam/compare/8dee1ec...15a5398
[0.2.1]: https://github.com/clitic/kdam/compare/80e2ea0...8dee1ec
[0.2.0]: https://github.com/clitic/kdam/compare/2e500d0...80e2ea0
[0.1.7]: https://github.com/clitic/kdam/compare/212923c...2e500d0
[0.1.6]: https://github.com/clitic/kdam/compare/323c3fa...212923c
[0.1.2]: https://github.com/clitic/kdam/compare/3f910c3...323c3fa
[0.1.0]: https://github.com/clitic/kdam/compare/58b20a4...3f910c3


src/term.rs
//! Terminal related functions for printing and text colorization.

use std::io::Write;
use std::sync::atomic::{AtomicBool, Ordering};

use unicode_segmentation::UnicodeSegmentation;

static COLOURS_ENABLED: AtomicBool = AtomicBool::new(false);

/// Stderr and Stdout writer for [Bar](crate::Bar).
#[derive(Debug, Clone)]
pub enum Writer {
    Stderr,
    Stdout,
}

impl From<&str> for Writer {
    fn from(output: &str) -> Self {
        match output.to_lowercase().as_str() {
            "stdout" => Self::Stdout,
            _ => Self::Stderr,
        }
    }
}

impl Writer {
    /// Print [Arguments](std::fmt::Arguments) in terminal followed by a flush.
    pub fn print(&self, args: std::fmt::Arguments) {
        match self {
            Self::Stderr => {
                let mut writer = std::io::stderr();
                writer.write_fmt(args).unwrap();
                writer.flush().unwrap();
            }
            Self::Stdout => {
                let mut writer = std::io::stdout();
                writer.write_fmt(args).unwrap();
                writer.flush().unwrap();
            }
        }
    }

    /// Print `&str` in terminal followed by a flush.
    pub fn print_str(&self, text: &str) {
        match self {
            Self::Stderr => {
                let mut writer = std::io::stderr();
                writer.write(text.as_bytes()).unwrap();
                writer.flush().unwrap();
            }
            Self::Stdout => {
                let mut writer = std::io::stdout();
                writer.write(text.as_bytes()).unwrap();
                writer.flush().unwrap();
            }
        }
    }

    /// Prints to the standard error at specified position.
    ///
    /// Also cursor position is restored to original position after print.
    ///
    /// # Example
    ///
    /// ```rust
    /// use kdam::term::Writer;
    ///
    /// Writer::Stderr.print_at(1, format!("1 + 1 = {}", 2));
    /// ```
    pub fn print_at<T: Into<String>>(&self, position: usize, text: T) {
        match self {
            Self::Stderr => {
                let mut writer = std::io::stderr();

                crate::lock::acquire();

                if position > 0 {
                    writer
                        .write_fmt(format_args!(
                            "{}{}\x1b[{}A",
                            "\n".repeat(position),
                            text.into(),
                            position
                        ))
                        .unwrap();
                } else {
                    writer.write_fmt(format_args!("{}", text.into())).unwrap();
                }

                writer.flush().unwrap();
                crate::lock::release();
            }
            Self::Stdout => {
                let mut writer = std::io::stdout();

                crate::lock::acquire();

                if position > 0 {
                    writer
                        .write_fmt(format_args!(
                            "{}{}\x1b[{}A",
                            "\n".repeat(position),
                            text.into(),
                            position
                        ))
                        .unwrap();
                } else {
                    writer.write_fmt(format_args!("{}", text.into())).unwrap();
                }

                writer.flush().unwrap();
                crate::lock::release();
            }
        }
    }
}

/// Get number of columns in current window or default to specified value.
pub fn get_columns_or(width: u16) -> u16 {
    terminal_size::terminal_size()
        .unwrap_or((terminal_size::Width(width), terminal_size::Height(0)))
        .0
         .0
}

/// Get display length of string.
pub fn string_display_length(mut text: String) -> usize {
    text = text.replace("\x1b[0m", "");

    while let Some(start) = text.find("\x1b[") {
        text = text.replace(
            &text[start..(start + text[start..].find("m").unwrap() + 1)],
            "",
        );
    }

    text.graphemes(true).count()
}

/// Create colour escape code from primary colours or hex colour code.
///
/// # Example
///
/// ```rust
/// use kdam::term::colour;
///
/// assert_eq!(colour("bold red"), "\x1b[31;1m");
/// assert_eq!(colour("blue on white"), "\x1b[34;47m");
/// ```
pub fn colour(colour_code: &str) -> String {
    if !COLOURS_ENABLED.load(Ordering::Acquire) {
        #[cfg(target_os = "windows")]
        std::process::Command::new("cmd")
            .args(["/c", "color"])
            .spawn()
            .unwrap()
            .wait()
            .unwrap();

        COLOURS_ENABLED.store(true, Ordering::SeqCst);
    }

    let mut color = colour_code.to_uppercase();
    let mut code = "\x1b[".to_string();

    let bg = if color.contains("ON BLACK") {
        color = color.replace("ON BLACK", "");
        ";40"
    } else if color.contains("ON RED") {
        color = color.replace("ON RED", "");
        ";41"
    } else if color.contains("ON GREEN") {
        color = color.replace("ON GREEN", "");
        ";42"
    } else if color.contains("ON YELLOW") {
        color = color.replace("ON YELLOW", "");
        ";43"
    } else if color.contains("ON BLUE") {
        color = color.replace("ON BLUE", "");
        ";44"
    } else if color.contains("ON MAGENTA") {
        color = color.replace("ON MAGENTA", "");
        ";45"
    } else if color.contains("ON CYAN") {
        color = color.replace("ON CYAN", "");
        ";46"
    } else if color.contains("ON WHITE") {
        color = color.replace("ON WHITE", "");
        ";47"
    } else {
        ""
    };

    // rgb(175,0,255)
    if let Some(hex_code) = colour_code.find("#") {
        code += &format!(
            "38;2;{};{};{}",
            i16::from_str_radix(&colour_code[(hex_code + 1)..(hex_code + 3)], 16).unwrap(),
            i16::from_str_radix(&colour_code[(hex_code + 3)..(hex_code + 5)], 16).unwrap(),
            i16::from_str_radix(&colour_code[(hex_code + 5)..(hex_code + 7)], 16).unwrap()
        );
    } else {
        if color.contains("BRIGHT") {
            if color.contains("BLACK") {
                code += "90";
            } else if color.contains("RED") {
                code += "91";
            } else if color.contains("GREEN") {
                code += "92";
            } else if color.contains("YELLOW") {
                code += "93";
            } else if color.contains("BLUE") {
                code += "94";
            } else if color.contains("MAGENTA") {
                code += "95";
            } else if color.contains("CYAN") {
                code += "96";
            } else if color.contains("WHITE") {
                code += "97";
            } else {
                return "".to_string();
            }
        } else {
            if color.contains("BLACK") {
                code += "30";
            } else if color.contains("RED") {
                code += "31";
            } else if color.contains("GREEN") {
                code += "32";
            } else if color.contains("YELLOW") {
                code += "33";
            } else if color.contains("BLUE") {
                code += "34";
            } else if color.contains("MAGENTA") {
                code += "35";
            } else if color.contains("CYAN") {
                code += "36";
            } else if color.contains("WHITE") {
                code += "37";
            } else {
                return "".to_string();
            }
        }
    }

    code += bg;

    if color.contains("BOLD") {
        code += ";1";
    }

    if color.contains("DIM") {
        code += ";2";
    }

    if color.contains("ITALIC") {
        code += ";3";
    }

    if color.contains("UNDERLINE") {
        code += ";4";
    }

    if color.contains("BLINK") {
        code += ";5";
    }

    if color.contains("REVERSED") {
        code += ";7";
    }

    if color.contains("HIDDEN") {
        code += ";8";
    }

    if color.contains("STRIKETHROUGH") {
        code += ";9";
    }

    code += "m";
    code
}

/// Add colour escape codes to the given text for printing coloured text in terminal.
/// This trait is only implemented for `&str`.
pub trait Colorizer {
    /// Add colour escape codes to the given text.
    ///
    /// # Example
    ///
    /// ```rust
    /// use kdam::prelude::*;
    ///
    /// println!("{}", "hello world!".colorize("bold red"));
    /// println!("{}", "hello world!".colorize("bright white on blue"));
    /// ```
    fn colorize<'a>(&'a self, code: &str) -> String;
}

impl Colorizer for str {
    fn colorize(&self, code: &str) -> String {
        let esc_code = colour(code);

        if esc_code == "" {
            self.to_owned()
        } else {
            esc_code + self + "\x1b[0m"
        }
    }
}
